src/annot/spec.ml

```diff
@@ -0,0 +1,378 @@
+  | BinPred : binpred * 'a term * 'a term -> 'a prop
+  | UnConnect : unconnect * 'a prop -> 'a prop
+  | BinConnect : binconnect * 'a prop * 'a prop -> 'a prop
+  | Binder : binder * binder_type * string option * 'a prop -> 'a prop
+
+let pp_bool fmt = function true -> pf fmt "true" | false -> pf fmt "false"
+
+let pp_binpred fmt = function
+  | Ge -> pf fmt "≥"
+  | Gt -> pf fmt ">"
+  | Le -> pf fmt "≤"
+  | Lt -> pf fmt "<"
+  | Eq -> pf fmt "="
+  | Neq -> pf fmt "≠"
+
+let pp_unconnect fmt = function Not -> pf fmt "¬"
+
+let pp_binconnect fmt = function
+  | And -> pf fmt "∧"
+  | Or -> pf fmt "∨"
+  | Imply -> pf fmt "⇒"
+  | Equiv -> pf fmt "⇔"
+
+let pp_binder fmt = function Forall -> pf fmt "∀" | Exists -> pf fmt "∃"
+
+let pp_binder_type = pp_num_type
+
+let pp_unop fmt = function
+  | Neg -> pf fmt "-"
+  | CustomUnOp c -> pf fmt "%a" string c
+
+let pp_binop fmt = function
+  | Plus -> pf fmt "+"
+  | Minus -> pf fmt "-"
+  | Mult -> pf fmt "*"
+  | Div -> pf fmt "/"
+  | CustomBinOp c -> pf fmt "%a" string c
+
+let rec pp_term : type a. formatter -> a term -> unit =
+ fun fmt -> function
+  | Int32 i32 -> pf fmt "(i32 %i)" (Int32.to_int i32)
+  | Int64 i64 -> pf fmt "(i64 %i)" (Int64.to_int i64)
+  | Float32 f32 -> pf fmt "(f32 %a)" Float32.pp f32
+  | Float64 f64 -> pf fmt "(f64 %a)" Float64.pp f64
+  | Var ind -> pf fmt "%a" pp_indice ind
+  | ParamVar ind -> pf fmt "(param %a)" pp_indice ind
+  | GlobalVar ind -> pf fmt "(global %a)" pp_indice ind
+  | BinderVar ind -> pf fmt "(binder %a)" pp_indice ind
+  | UnOp (u, tm1) -> pf fmt "@[<hv 2>(%a@ %a)@]" pp_unop u pp_term tm1
+  | BinOp (b, tm1, tm2) ->
+    pf fmt "@[<hv 2>(%a@ %a@ %a)@]" pp_binop b pp_term tm1 pp_term tm2
+  | Result -> pf fmt "result"
+
+let rec pp_prop : type a. formatter -> a prop -> unit =
+ fun fmt -> function
+  | Const bool -> pf fmt "%a" pp_bool bool
+  | BinPred (b, tm1, tm2) ->
+    pf fmt "@[<hv 2>(%a@ %a@ %a)@]" pp_binpred b pp_term tm1 pp_term tm2
+  | UnConnect (u, pr1) -> pf fmt "@[<hv 2>(%a@ %a)@]" pp_unconnect u pp_prop pr1
+  | BinConnect (b, pr1, pr2) ->
+    pf fmt "@[<hv 2>(%a@ %a@ %a)@]" pp_binconnect b pp_prop pr1 pp_prop pr2
+  | Binder (b, bt, id_opt, pr1) -> (
+    match id_opt with
+    | Some id ->
+      pf fmt "@[<hv 2>(%a %a:%a@ %a)@]" pp_binder b pp_id id pp_binder_type bt
+        pp_prop pr1
+    | None ->
+      pf fmt "@[<hv 2>(%a %a@ %a)@]" pp_binder b pp_binder_type bt pp_prop pr1 )
+
+let valid_text_indice_char = function
+  | '0' .. '9'
+  | 'a' .. 'z'
+  | 'A' .. 'Z'
+  | '!' | '#' | '$' | '%' | '&' | '\'' | '*' | '+' | '-' | '.' | '/' | ':' | '<'
+  | '=' | '>' | '?' | '@' | '\\' | '^' | '_' | '`' | '|' | '~' ->
+    true
+  | _ -> false
+
+let parse_text_id id =
+  try
+    let len = String.length id in
+    let hd = String.get id 0 in
+    let tl = String.sub id 1 (len - 1) in
+    if Char.equal hd '$' && String.for_all valid_text_indice_char id then
+      Some tl
+    else None
+  with Invalid_argument _ -> None
```

Comment:
You should never catch `Invalid_argument _` when writing OCaml code. `Invalid_argument` denotes a programming error and you should ensure before calling the function that it won't happen. This is really needed because when compiling with `-unsafe`, such exception is not going to be raised and the code will segfault or return a garbage value...

Reply:
Ok thanks!